File: app/src/main/java/io/github/lazymockbean/utils/MockGenerator.java
package io.github.lazymockbean.utils;

import io.github.lazymockbean.annotation.LazySpyBean;
import lombok.experimental.UtilityClass;
import org.mockito.AdditionalAnswers;
import org.mockito.Mockito;
import org.springframework.aop.support.AopUtils;
import org.springframework.test.context.TestContext;

import java.lang.reflect.Field;

@UtilityClass
public class MockGenerator {
    public Object generate(Field field, TestContext testContext, Class<?> targetFieldType) {
        if (field.isAnnotationPresent(LazySpyBean.class)) {
            return getSpyMock(testContext, targetFieldType);
        }
        return Mockito.mock(targetFieldType);
    }

    private Object getSpyMock(TestContext testContext, Class<?> targetFieldType) {
        Object spyTargetBean = SpringBeans.findBean(testContext, targetFieldType);
        if (AopUtils.isAopProxy(spyTargetBean)) {
            return Mockito.mock(targetFieldType, AdditionalAnswers.delegatesTo(spyTargetBean));
        } else {
            return Mockito.spy(spyTargetBean);
        }
    }
}

File: app/src/main/kotlin/io/github/taesk/lazymockbean/utils/MockGenerator.kt
package io.github.taesk.lazymockbean.utils

import io.github.taesk.lazymockbean.annotation.LazySpyBean
import org.mockito.AdditionalAnswers
import org.mockito.Mockito
import org.springframework.aop.support.AopUtils
import org.springframework.test.context.TestContext
import java.lang.reflect.Field

object MockGenerator {
    fun generate(field: Field, testContext: TestContext, targetFieldType: Class<*>): Any {
        if (field.isAnnotationPresent(LazySpyBean::class.java)) {
            return getSpyMock(testContext, targetFieldType)
        }
        return Mockito.mock(targetFieldType)
    }

    private fun getSpyMock(testContext: TestContext, targetFieldType: Class<*>): Any {
        val spyTargetBean = SpringBeans.findBean(testContext, targetFieldType)
        return if (AopUtils.isAopProxy(spyTargetBean)) {
            Mockito.mock(targetFieldType, AdditionalAnswers.delegatesTo<Any>(spyTargetBean))
        } else {
            Mockito.spy(spyTargetBean)
        }
    }
}
